### PR TITLE
[PExplicit] Support liveness/deadlock checking, state-caching, stateful backtracking

### PR DESCRIPTION
**Pull request by aman-goel**
Updates include:

[Deadlock/liveness]
- Adds deadlock checking
- Adds basic liveness checking, based on whether a monitor is in a hot state at the end of a schedule

[State caching]
- Adds state caching to avoid revisiting same state more than once
- State caching mode can be: `none`, `fingerprint` (i.e., use hash codes), or `exact` (use exact values). Fingerprinting is the default.
- Adds hidden CLI option `--state-caching <mode>` (default is `fingerprint`)

[Stateful backtracking]
- Adds stateful backtracking to directly jump to the choice state instead of restarting from step zero at every iteration.
- Each unexplored choice stores the scheduler state at the start of the corresponding scheduler step.
- Stateful backtracking is completely optional and can be later tuned. If the unexplored choice selected in a new iteration has the stored state, stateful backtracking is done. If stored state is null, restarting from step zero is done.
- Adds hidden CLI option `--no-backtrack` to disable stateful backtracking (by default it is enabled).

[Replaying]
- Adds buggy trace replayer to confirm and replay the buggy trace when a bug is found.
- Adds basic `.schedule` writer when replaying (currently based on sender queue semantics)
- Adds TextWriter to write trace in `.txt` format when replaying (currently based on sender queue semantics)

[Misc]
- Updates logger to resemble like in C#
- Updates logging and stats to track and report distinct states
- Minor fixes, reformatting, and corrections

### CODE REVIEW

File: Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/runtime/machine/MachineLocalState.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * Represents the local state of a machine
+ */
+public class MachineLocalState implements Serializable {
+    /**
+     * List of values of all local variables (including internal variables like currentState, FIFO queue, etc.)
```
Comment (ankushdesai) [inline Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/runtime/machine/MachineLocalState.java:14]: I am a bit worried if this is the best way to track local state.

File: Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/values/ComputeHash.java [hunk 2]
```diff
@@ -43,4 +45,18 @@ public static int getHashCode(PMachine machine, PValue<?>... values) {
         }
         return hashValue;
     }
+
+    /**
+     * Compute hash value for a PMachine local variables.
+     */
+    public static int getHashCode(SortedSet<PMachine> machines) {
```
Comment (ankushdesai) [inline Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/values/ComputeHash.java:52]: I am hoping this way of computing hash is prototype and we improve on this?

File: Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/runtime/machine/PMachine.java [hunk 3]
```diff
@@ -125,29 +120,122 @@ public boolean canRun() {
...
+        return result;
     }
 
     /**
-     * TODO
+     * Get values of local variables as an ordered list
+     *
+     * @return List of values
+     */
+    public List<Object> getLocalVarValues() {
```
[the remark this answers is not in the document]
Comment (aman-goel) [reply Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/runtime/machine/PMachine.java:165]: This is the first version. As we develop this further, we can update it.

For state caching and stateful backtracking, we need a way to access all field values associated with the machine. These include all field values declared as the subclass that extends the `PMachine` abstract class (declared in corresponding PExplicit IR) as well as all fields inherited from the `PMachine`.

An alternative is to use [getDeclaredFields()](https://stackoverflow.com/questions/2172263/java-reflection-find-fields-of-a-subclass) Java function, but things become more complicated with monitors, making the inheritance to for e.g., `PMachine` -> `PMonitor` -> `Client`.

Any suggestions welcomed!

File: Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/runtime/machine/MachineLocalState.java (hunk 1 above)
Comment (aman-goel) [reply Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/runtime/machine/MachineLocalState.java:14]: Let's discuss this further offline and revamp this in the next PR.

File: Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/values/ComputeHash.java (hunk 2 above)
Comment (aman-goel) [reply Src/PRuntimes/PExplicitRuntime/src/main/java/pexplicit/values/ComputeHash.java:52]: Yes, indeed. A much improved version of state caching to follow in the next PR.

**Review by ankushdesai — APPROVED** [review]

**Pull request merged**